x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,329 @@
+        );
+        return false;
+      }
+
+      this.logger.debug(`Lock "${this.lockId}" with token = ${this.token} was released.`);
+      return true;
+    } catch (error) {
+      this.logger.error(`Failed to release lock "${this.lockId}": ${error.message}`);
+      this.logger.debug(error);
+      return false;
+    }
+  }
+
+  /**
+   * Releases a lock if it is expired.
+   */
+  public async releaseIfExpired(): Promise<boolean> {
+    const res = await this.esClient.deleteByQuery({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { lte: 'now' } } }],
+        },
+      },
+      refresh: true,
+    });
+    if (res.deleted === 0) {
+      this.logger.debug(`Lock "${this.lockId}" not found`);
+      return false;
+    }
+
+    this.logger.debug(`Lock "${this.lockId}" expired and was released.`);
+    return true;
+  }
+
+  /**
+   * Retrieves the lock document for a given lockId.
+   * If the lock is expired, it will be released.
+   */
+  public async get(): Promise<LockDocument | undefined> {
+    const result = await this.esClient.search<LockDocument>({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { gt: 'now' } } }],
+        },
+      },
+    });
+
+    const hits = result.hits.hits;
+    if (hits.length === 0) {
+      await this.releaseIfExpired();
+      return undefined;
+    }
+
+    return hits[0]._source;
+  }
+
+  public async waitForLock(options: AcquireOptions & pRetry.Options = {}): Promise<boolean> {
+    const { metadata, ttlMs, ...pRetryOptions } = options;
+
+    return pRetry(async () => {
+      const acquired = await this.acquire({ metadata, ttlMs });
+      if (!acquired) {
+        this.logger.debug(`Lock "${this.lockId}" not available yet.`);
+        throw new Error(`Lock "${this.lockId}" not available yet`);
+      }
+      return acquired;
+    }, pRetryOptions ?? { forever: true, maxTimeout: 10_000 });
+  }
+
+  public async extendTtl(ttl = 300000): Promise<boolean> {
+    try {
+      await this.esClient.update<LockDocument>({
+        index: LOCKS_CONCRETE_INDEX_NAME,
+        id: this.lockId,
+        script: {
+          lang: 'painless',
+          source: `
+          long now = System.currentTimeMillis();
+          ctx._source.expiresAt = Instant.ofEpochMilli(now + params.ttl).toString();`,
+          params: { ttl },
+        },
+        refresh: true,
+      });
+      this.logger.debug(`Lock "${this.lockId}" extended ttl with ${prettyMilliseconds(ttl)}.`);
+      return true;
+    } catch (error) {
+      if (
+        error instanceof errors.ResponseError &&
+        error.message.includes('version_conflict_engine_exception') &&
+        error.message.includes('but no document was found')
```

Comment:
what is this about?

Reply:
Right, isn't `version_conflict_engine_exception` enough to check this?

Reply:
makes sense that we check for `version_conflict_engine_exception`, but do we need anything else?

Reply:
No, you are probably right. I'll reduce the check to just `version_conflict_engine_exception `